nixos/k3s: add initial k3s service

euank opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

I thought that others might find this service definition handy! I'll use it whether we land it in the tree or not.

###### Things done

I added a k3s service with a few options that I personally find handy.

Note that this is my first nixos module, so it's _very_ likely I'm missing something I should do. I don't know if I need to do more for docs, or if there's more testing I need to do.

I would definitely appreciate any pointers for stuff I might have missed or should do for a new module. If there's an exemplary pull request I could mimic, I'd be happy to take inspiration from that.

###### Testing done
I tested this on my nixos machine with the following config:

```
  services.k3s.enable = true;
  services.k3s.role = "server";
  services.k3s.docker = true;
```

With that configuration, running `sudo k3s kubectl get pod` and such worked as I'd expect.

euank committed:
nixos/k3s: add initial k3s service

ngerstle commented:
Looks very similar to what I have, nice work!

ngerstle commented on nixos/modules/services/cluster/k3s/default.nix:
```diff
@@ -0,0 +1,104 @@
+              TasksMax = "infinity";
+              Restart = "always";
+              RestartSec = "5s";
+              ExecStart = "${pkgs.k3s}/bin/k3s ${cfg.role} ${flags}";
```
perhaps allow specifying an alternative package?
```
 package = mkOption {
      default = pkgs.k3s;
      defaultText = "pkgs.k3s";
      type = types.package;
      description = "k3s package to use.";
    };
```
above with 
```
 ExecStart = "${cfg.package.bin}/bin/k3s ${cfg.role} ${flags}";
```
here instead..
```

euank replied:
I think I plumbed it through correctly, thanks for the suggestion

ngerstle reviewed (approved):
Would like to see a package option exposed, but LGTM!

misuzu commented on nixos/modules/services/cluster/k3s/default.nix:
```diff
@@ -0,0 +1,104 @@
+            };
+          };
+
+        environment.systemPackages = [ pkgs.k3s ] ++ (if cfg.docker then [ pkgs.docker ] else []);
```
`++ (if cfg.docker then [ pkgs.docker ] else [])` part is [redundant](https://github.com/NixOS/nixpkgs/blob/47daa683c9f38f9ae84a3d1cd6f50583215ad0c0/nixos/modules/virtualisation/docker.nix#L152).

euank replied:
Ah, yup, makes sense

misuzu commented on nixos/modules/services/cluster/k3s/default.nix:
```diff
@@ -0,0 +1,104 @@
+    (
+      mkIf cfg.docker {
+        virtualisation.docker = {
+          enable = mkDefault true;
```
This would be better:
```
virtualisation.docker = mkIf cfg.docker {
  enable = mkDefault true;
}
```

misuzu commented on nixos/modules/services/cluster/k3s/default.nix:
```diff
@@ -0,0 +1,104 @@
+    (
+      mkIf (cfg.role == "server") {
+        services.k3s.serverAddr = "";
+        services.k3s.token = "";
```
Use this and `mkMerge` can be dropped.
```
services.k3s = mkIf (cfg.role == "server") {
  serverAddr = "";
  token = "";
};
```

euank replied:
Thanks for that suggestion! That's definitely cleaner looking.

misuzu commented on nixos/modules/services/cluster/k3s/default.nix:
```diff
@@ -0,0 +1,104 @@
+    (
+      mkIf cfg.enable {
+        systemd.services.k3s = let
+          flags = concatStrings (
```
Take a look at similar piece in [`docker-container`](https://github.com/NixOS/nixpkgs/blob/47daa683c9f38f9ae84a3d1cd6f50583215ad0c0/nixos/modules/virtualisation/docker-containers.nix#L224) module.

misuzu commented on nixos/modules/services/cluster/k3s/default.nix:
```diff
@@ -0,0 +1,104 @@
+              remove "" [
+                (if cfg.docker then "--docker" else "")
+                (if (cfg.serverAddr != "") then "--server ${cfg.serverAddr}" else "")
+                (if (cfg.token != "") then "--token ${cfg.token}" else "")
```
Here could be something like this:
```
optional (cfg.role == "server") "--server ${cfg.serverAddr} --token ${cfg.token}"
```
So setting `serverAddr` and `token` to empty string will not be required. Default value for `serverAddr` could also be dropped.

misuzu commented:
It might make sense to separate server and agent to corresponding modules and systemd units.
```
{
  services.k3s = {
    agent = {
      enable = true;
      docker = true;
    };
    server.enable = true;
  };
}
```

euank committed:
nixos/k3s: simplify config expression

euank commented:
Thanks for all the help cleaning this up @misuzu! I think I made all the improvements you suggested, and it looks quite a bit better for it.

On whether we should split agent and server, I'm not totally sure the right thing to do there.

I do think it makes more sense to keep it as one systemd service at least. The upstream documentation / installer assumes it's "k3s.service", so it makes it less surprising if "systemctl status k3s.service" etc works on nixos too.
There's also very little reason to want to run both services independently (i.e. the server role by default already runs the agent, so if you want both, you've already got both).

Because of the above, I ended up with the single config structure and single module.

nlewo commented on nixos/modules/services/cluster/k3s/default.nix:
```diff
@@ -0,0 +1,100 @@
+    disableAgent = mkOption {
+      type = types.bool;
+      default = false;
+      description = "Only run the server. This option only makes sense for a server.";
```
Ideally, you could add a `assert` in the config to check this.

nlewo commented on nixos/modules/services/cluster/k3s/default.nix:
```diff
@@ -0,0 +1,100 @@
+        Type = "notify";
+        KillMode = "process";
+        Delegate = "yes";
+        LimitNOFILE = "infinity";
```
Not sure we want infinity everywhere. Is it required to make k3s working? A user really wanting this could still override these values.
So, if this is not strictly required, could you remove all these `infinity` values?

euank replied:
It's not strictly required, but it's _often_ what you want I think.

If you're using the builtin containerd runtime in k3s, you'll hit nofile limits pretty quickly if you run a lot of pods. I removed it for now since I'm not totally confident in this.

nlewo commented on nixos/modules/services/cluster/k3s/default.nix:
```diff
@@ -0,0 +1,100 @@
+      };
+    };
+
+    environment.systemPackages = [ cfg.package ];
```
I don't think we have to force users having k3s globally installed.

nlewo reviewed (changes requested):
Thanks for you contribution! This is a nice first module;)

Could you add a NixOS test for this module? You could for instance look at the docker test implementation: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/docker.nix

euank committed:
nixos/k3s: add config assertions and trim unneeded bits

euank committed:
nixos/k3s: add a test that k3s works; minor module improvements

This is a single-node test. Eventually we should also have a multi-node
test to verify the agent bit works, but that one's more involved.

offlinehacker commented on nixos/modules/services/cluster/k3s/default.nix:
```diff
@@ -0,0 +1,105 @@
+      enable = mkDefault true;
+    };
+
+    services.k3s = mkIf (cfg.role == "server") {
```
You can put default to `mkOption`, which makes it more readable.

nlewo replied:
The current implementation has the advantage of forcing the user to set a `serverAddr` when the role is `agent`.

euank replied:
I flipped the assertion so it's now caught that way. I assume @offlinehacker meant that an assertion is preferable to a type-error here.
I can flip it back if we'd prefer one or the other.

nlewo replied:
No, that's fine with an assertion.

offlinehacker commented on nixos/modules/services/cluster/k3s/default.nix:
```diff
@@ -0,0 +1,105 @@
+      default = pkgs.k3s;
+      defaultText = "pkgs.k3s";
+      example = literalExample "pkgs.k3s";
+      description = ''
```
Missing description

offlinehacker reviewed (changes requested):
Just two small changes needed, but everything else seems ok. Thanks!

euank committed:
nixos/k3s: add option description

nlewo commented on nixos/tests/k3s.nix:
```diff
@@ -0,0 +1,80 @@
+    start_all()
+
+    k3s.wait_for_unit("k3s")
+    k3s.succeed("sudo k3s kubectl cluster-info")
```
I don't think `sudo` is required since tests are executed as root.

euank committed:
nixos/k3s: add defaults for token/serveraddr

Now that the assertion enforces their presence, we dont' need to use the typesystem for it.

euank commented:
I think I made the changes you requested @offlinehacker, and I also added a test (though I only ran it a couple times, so beware it may have some inconsistent flake or race).

euank committed:
nixos/k3s: remove unneeded sudo in test

nlewo commented:
@euank You also need to reference the test in `nixos/tests/all_tests.nix`.

euank committed:
nixos/k3s: add to test list

nlewo commented:
@GrahamcOfBorg test k3s

nlewo reviewed (approved):
Thanks for your contribution!
